src/PaymentServer/Ristretto.hs

```diff
@@ -85,6 +92,20 @@ data RistrettoFailure
   | PublicKeyEncoding
   deriving (Show, Eq)
 
+nullIsError :: b -> IO (Ptr a) -> ExceptT b IO (Ptr a)
+nullIsError fallback op =
+  liftIO op >>= \r ->
+                  case r of
+                    nullPtr -> throwError fallback
+                    ptr -> return ptr
+
+anyNullIsError :: b -> IO [Ptr a] -> ExceptT b IO [Ptr a]
+anyNullIsError fallback op =
+  liftIO op >>= \r ->
+                  if nullPtr `elem` r
+                  then throwError fallback
+                  else return r
+
 ristretto
   :: Text                                  -- ^ The base64 encoded signing key.
   -> [Text]                                -- ^ A list of the base64 blinded tokens.
```

Comment:
Any reason against having the return type be `IO [Text]` ?
That would remove the `unsafePerformIO` but I don't know the codebase well enough to know if there's a good reason for this.

Reply:
Whatever the original reason was, I can't remember it.  It was _probably_ something spurious or transient like "but I don't how to write a unit test for an `IO m` function".  I would like for this mistake to be fixed in a follow-up PR.